### PR TITLE
Add distinction between line-averaged and volume-averaged density

### PR DESCRIPTION
**Pull request by nelsonand**
A pull request to organize solutions to https://github.com/cfs-energy-internal/internal-cfspopcon/issues/237 and https://github.com/cfs-energy-internal/internal-cfspopcon/issues/238. Most of this development was done by Columbia undergrads @aarooshgr and @ThomasWangAPAM

### CODE REVIEW

File: example_cases/SPARC_PRD/input.yaml
Comment (nelsonand) [inline example_cases/SPARC_PRD/input.yaml:134]: This needs to have a better name...

File: cfspopcon/unit_handling/default_units.py
Comment (nelsonand) [inline cfspopcon/unit_handling/default_units.py:49]: Need to update name here

File: cfspopcon/algorithms/peaked_profiles.py
Comment (nelsonand) [inline cfspopcon/algorithms/peaked_profiles.py:69]: This should also be `line_averaged_electron_density`

**Comment by nelsonand** [comment]
**Summary of changes**

Two average densities now exist:
- The volume-averaged density (`<ne>`, `average_electron_density`) 
- The line-averaged density (`ne_bar`, `line_average_electron_density`)

Ion and electron densities are now handled separately when needed. 

These changes were propagated through the code as specified in [#238](https://github.com/cfs-energy-internal/internal-cfspopcon/issues/238): 

- [ ] Add a function that relates `ne_bar` to `<ne>`
- [x] Add a definition of `ne_bar` to physics glossary
- [ ] Use `ne_bar` in confinement calcs, profile calcs and upstream density calcs
- [x] Beta definitions should use the one instance of calc_average_total_pressure to be consistent

1. Do the PRF profiles use <ne> or ne_bar?  The name n_avol suggests a volume average.
2. PBpRnSq … heat exhaust … which definition of density?
3. Two-point models … which definition of density?
4. Mean charge curve for radiation … which definition of density?

Full list follows:

Calc_average_total_pressure
- Uses `<ne>` This is correct
- [x] Now updated to include ion and electron density contributions separately

Calc_beta_general
- [x] Should use average_total_pressure instead of recomputing the average_total_pressure

Run_calc_beta, calc_beta_toroidal, calc_beta_poloidal
- [x] Same as calc_beta_general, use average_total_pressure instead of recomputing

Two_point_model_fixed_fpow
- Computes upstream electron density using `nesep_over_nebar`
- [x] This should use `ne_bar` instead of `<ne>`

Physics_glossary
- [ ] Define aveage_electron_density with the symbol <ne>
- [x] Add definition of line average density ne_bar

use_LOC_tau_e_below_threshold
- energy_confinement_time_LOC
- [x] Should use `ne_bar`

Figures_of_merit
- Calc_normalized_collisionality
- OK to use `<ne>` 

Power_balance_from_tau, Run_calc_power_balance_from_tau_e
- [x] Energy_confinement_time should use`ne_bar`

Operational_limits
- [x] Calc_greenwald_fraction should use ne_bar

Heat exhaust
- run_calc_heat_exhaust
-  PBpRnSq = (P_sol * (magnetic_field_on_axis / q_star) / major_radius) / (average_electron_density**2.0)
- **Not sure, left as `<ne>`** 

Plasma_profiles
- calc_1D_plasma_profiles
- [x] calc_analytic_profiles should use `ne_bar` 
- calc_prf_profiles uses  `<ne>` accroding to @pabloprf 
- Evaluate_density_and_temperature_profile_fits

Peaked_profiles
- Run_calc_peaked_profiles
- Calc_effective_collisionality: ok to `<ne>`
- Peak_electron_density … use`<ne>`
- [ ] calc_1D_plasma_profiles … use ne_bar

Two_point_model_fixed_tet
- Run_two_point_model_fixed_tet
- **Not sure, left as `<ne>`** 

Two_point_model_fixed_qpart
- **Not sure, left as `<ne>`** 

Extrinsic_core_radiator
- Run_calc_extrinsic_core_radiator
- Calc_impurity_charge_state
- mean_charge_curve
- **Not sure, left as `<ne>`** 

Zeff_and_dilution_from_impurities
Run_calc_zeff_and_dilution_from_impurities
Calc_impurity_charge_state
Summed_impurity_density
- Use `<ne>`
- [x] More correct to call this summed_impurity_average_density

Average_ion_density
- Use `<ne>`

**Comment by nelsonand** [comment]
@hassec @djbattagl Should be ready for a review. 

This addresses all of https://github.com/cfs-energy-internal/internal-cfspopcon/issues/238, except for updating the symbol as indicated in that issue (I wasn't sure if we want that or not).

https://github.com/cfs-energy-internal/internal-cfspopcon/issues/238 does have a few things as "not sure" - we didn't change these in the end.

File: docs/doc_sources/physics_glossary.rst
Comment (djbattagl) [reply docs/doc_sources/physics_glossary.rst:200]: I don't think it is necessary to change this definition, unless you think it makes it clearer or more consistent. @nelsonand

**Comment by djbattagl** [comment]
@tbody-cfs Do you have any quick guidance on whether to use volume averaged density or line-averaged density in these modules: run_calc_heat_exhaust, Run_two_point_model_fixed_tet, Two_point_model_fixed_qpart, Extrinsic_core_radiator ?

**Comment by djbattagl** [comment]
@nelsonand The code looks good to me. Could you provide an example where line_averaged_density_frac = 1 and compare to the results from an older version (should be the same). 

And then show how the results change with the fraction not equal to one.

File: docs/doc_sources/physics_glossary.rst
Comment (tbody-cfs) [reply docs/doc_sources/physics_glossary.rst:200]: It would make sense to call this `volume_averaged_electron_density`

Comment (nelsonand) [reply docs/doc_sources/physics_glossary.rst:200]: I agree with @tbody-cfs about this in principle, but for consistency we would have to propagate `volume_averaged_electron_temperature`, etc. through the code as well, which seems like it would just add more clutter everywhere. 

Can do if desired!

**Comment by tbody-cfs** [comment]
> @tbody-cfs Do you have any quick guidance on whether to use volume averaged density or line-averaged density in these modules: run_calc_heat_exhaust, Run_two_point_model_fixed_tet, Two_point_model_fixed_qpart, Extrinsic_core_radiator ?

I'd use volume-averaged for these. It really only depends on your choice of `nesep_over_nebar = 0.3`.

File: cfspopcon/algorithms/peaked_profiles.py
Comment (hassec) [inline cfspopcon/algorithms/peaked_profiles.py:70]: We should resolve this question.

File: cfspopcon/formulas/beta.py
Comment (hassec) [inline cfspopcon/formulas/beta.py:10]: Citation only really says `2*m_0 * <p> / B**2` so I'd say citation is still ok. and we only change how we calculate the average pressure. What do you think?

File: docs/doc_sources/physics_glossary.rst
Comment (hassec) [reply docs/doc_sources/physics_glossary.rst:200]: I'd say I'm ambivalent.  Having the variable called `volume_averaged_electron_temperature` would be nicer, but it is a tedious job to do the renaming and the glossary correctly explains that this is the volume averaged electron density.

So @nelsonand, IMHO it's ok as is.

File: poetry.lock
Comment (hassec) [inline poetry.lock:1]: This looks like somebody ran `poetry update`. 

Would you be ok with reverting this? I prefer doing dependency updates as standalone PR to be able to disentangle potential changes from updating from those in feature changes.

File: radas
Comment (hassec) [inline radas:1]: Why are we deleting `radas` here?

**Review by hassec — CHANGES_REQUESTED** [review]
Thanks, @nelsonand 😊 

Some small comments here and there. And we need to make sure the CI tests pass. 

Once that's working we would be ok from my side. But I'd like a 👍 from @djbattagl and @tbody-cfs on the physics changes.

File: cfspopcon/formulas/beta.py
Comment (nelsonand) [reply cfspopcon/formulas/beta.py:10]: I agree!

**Comment by nelsonand** [comment]
@hassec I'm having trouble getting radas to install on this branch (and the main `cfspopcon` branch) Are the installation instructions up to date after the latest updates?

File: radas
Comment (nelsonand) [reply radas:1]: We shouldn't be... I'm having trouble restoring this.

**Comment by hassec** [comment]
@nelsonand the way to install/use radas did indeed change but should be updated and working on `main`. 
That's a change that came in via the merge of #40. 

This branch seems to have quite a few conflicts (probably due to the merge of #40). 
So the best bet is likely a rebase of this branch, fixing the conflicts, and then trying again. 
That said there is also #41 which might get merged soon. I think you and @tbody-cfs already chatted about that PR, right? So it might make sense to wait for #41 and only rebase after that merge to avoid having to rebase and resolve conflicts twice.

**Comment by nelsonand** [comment]
Yeah - good call. I guess we'll wait for https://github.com/cfs-energy/cfspopcon/pull/41 to go in first! Thanks.

**Comment by nelsonand** [comment]
Hi @ThomasWangAPAM , can you rebase this onto https://github.com/cfs-energy/cfspopcon/pull/41 in prep for the merge?

**Comment by ThomasWangAPAM** [comment]
> Hi @ThomasWangAPAM , can you rebase this onto #41 in prep for the merge?

Rebase done!

**Comment by ThomasWangAPAM** [comment]
> @nelsonand The code looks good to me. Could you provide an example where line_averaged_density_frac = 1 and compare to the results from an older version (should be the same).
> 
> And then show how the results change with the fraction not equal to one.

POPCON generated by the current cfspopcon/main branch:
image

POPCON generated by new codes with `line_averaged_density_frac` = 1:
image

POPCON generated by new codes with `line_averaged_density_frac` = 1.5:
image

**Comment by hassec** [comment]
@nelsonand @ThomasWangAPAM what is the status of this PR. Is there a plan to rebase and finalize this?

**Comment by tbody-cfs** [comment]
@hassec I can do the rebase after #80 goes in, since this was already rebased once

**Pull request closed without merging**
